File: src/s0102_binary_tree_level_order_traversal.rs
/**
 * [102] Binary Tree Level Order Traversal
 *
 * Given a binary tree, return the level order traversal of its nodes' values. (ie, from left to right, level by level).
 *
 *
 * For example:
 * Given binary tree [3,9,20,null,null,15,7],
 *
 *     3
 *    / \
 *   9  20
 *     /  \
 *    15   7
 *
 *
 *
 * return its level order traversal as:
 *
 * [
 *   [3],
 *   [9,20],
 *   [15,7]
 * ]
 *
 *
 */

pub struct Solution;

// problem: https://leetcode.cn/problems/binary-tree-level-order-traversal/

// submission codes start here
use std::cell::RefCell;
use std::collections::VecDeque;
use std::rc::Rc;
use crate::tree;

type TreeNode = crate::TreeNode<i32>;

impl Solution {
    pub fn level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        if root.is_none() { return vec![]; }
        let mut ret = Vec::new();
        let mut queue = VecDeque::new();

        queue.push_back(root);
        while !queue.is_empty() {
            let mut level = vec![];
            let size = queue.len();
            for _ in 0..size {
                if let Some(maybe_n) = queue.pop_front() {
                    match maybe_n {
                        Some(n) => {
                            level.push(n.borrow().val);
                            queue.push_back(n.borrow().left.clone());
                            queue.push_back(n.borrow().right.clone());
                        },
                        None => ()
                    }
                }
            }
            if !level.is_empty() { ret.push(level); }
        }

        ret
    }
}

#[test]
fn test_102() {
    assert_eq!(
        Solution::level_order(tree![3, 9, 20, null, null, 15, 7]),
        vec![vec![3], vec![9, 20], vec![15, 7]]
    );
}

File: cpp/CPP/violent-search/s0102_binary_tree_level_order_traversal.cpp
// problem: https://leetcode.cn/problems/binary-tree-level-order-traversal/
#include "../data_structure.h"
#include <vector>
#include <queue>
using namespace std;

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> ret;
        if(root == nullptr) { return ret; }
        queue<TreeNode*> q;
        q.push(root);

        while(!q.empty()) {
            int sz = q.size();
            vector<int> level;
            for(int i = 0; i < sz; ++i) {
                TreeNode* n = q.front();
                q.pop();
                level.push_back(n->val);
                if(n->left) { q.push(n->left); }
                if(n->right) { q.push(n->right); }
            }
            if(!level.empty()) { ret.push_back(level); }
        }
        
        return ret;
    }
};

int main() { return 0; }